File: getPubKeyFromRSZ.js
var BN = require("big-integer");

var jordan_isinf = function (p) {
    return p[0][0].eq(ZERO) && p[1][0].eq(ZERO);
}

var mulcoords = function (c1, c2) {
    return [c1[0].multiply(c2[0]).mod(P), c1[1].multiply(c2[1]).mod(P)];
}

var mul_by_const = function (c, v) {
    return [c[0].multiply(v).mod(P), c[1]];
}

var addcoords = function (c1, c2) {
    return [c1[0].multiply(c2[1]).add(c2[0].multiply(c1[1])).mod(P), c1[1].multiply(c2[1]).mod(P)];
}

var subcoords = function (c1, c2) {
    var aa = c1[0].multiply(c2[1]).mod(P);
    var bb = c2[0].multiply(c1[1]).mod(P);
    var cc = aa.subtract(bb).add(P).mod(P);
    return [cc.mod(P), c1[1].multiply(c2[1]).mod(P)];
}

var invcoords = function (c) {
    return [c[1], c[0]];
}

var jordan_add = function (a, b) {
    if (jordan_isinf(a)) return b;
    if (jordan_isinf(b)) return a;

    if (a[0][0].multiply(b[0][1]).subtract(b[0][0].multiply(a[0][1])).mod(P).eq(0)) {
        if (a[1][0].multiply(b[1][1]).subtract(b[1][0].multiply(a[1][1])).mod(P).eq(0)) {
            return jordan_double(a);
        } else {
            return [[ZERO, ONE], [ZERO, ONE]];
        }
    }
    var xdiff = subcoords(b[0], a[0]);
    var ydiff = subcoords(b[1], a[1]);
    var m = mulcoords(ydiff, invcoords(xdiff));
    var x = subcoords(subcoords(mulcoords(m, m), a[0]), b[0]);
    var y = subcoords(mulcoords(m, subcoords(a[0], x)), a[1]);
    return [x, y]
}

var jordan_double = function (a) {
    if (jordan_isinf(a)) {
        return [[ZERO, ONE], [ZERO, ONE]];
    }
    var num = addcoords(mul_by_const(mulcoords(a[0], a[0]), THREE), [ZERO, ONE]);
    var den = mul_by_const(a[1], TWO);
    var m = mulcoords(num, invcoords(den));
    var x = subcoords(mulcoords(m, m), mul_by_const(a[0], TWO));
    var y = subcoords(mulcoords(m, subcoords(a[0], x)), a[1]);
    return [x, y]
}

var jordan_multiply = function (a, n) {
    if (jordan_isinf(a) || n.eq(ZERO)) {
        return ([ZERO, ZERO], [ZERO, ZERO]);
    }
    if (n.eq(1)) {
        return a;
    }
    if (n.lt(0) || n.geq(N)) {
        return jordan_multiply(a, n.mod(N));
    }
    if (n.mod(2).eq(ZERO)) {
        return jordan_double(jordan_multiply(a, n.divide(TWO)));
    }
    if (n.mod(2).eq(ONE)) {
        return jordan_add(jordan_double(jordan_multiply(a, n.divide(TWO))), a);
    }
}

var to_jordan = function (p) {
    return [[p[0], ONE], [p[1], ONE]];
}

var from_jordan = function (p) {
    return [p[0][0].multiply(p[0][1].modInv(P)).mod(P), p[1][0].multiply(p[1][1].modInv(P)).mod(P)];
}

var mul = function (a, n) {
    return from_jordan(jordan_multiply(to_jordan(a), n))
}

var div = function (a, n) {
    return from_jordan(jordan_multiply(to_jordan(a), n.modInv(P).mod(P)))
}

var add = function (a, b) {
    return from_jordan(jordan_add(to_jordan(a), to_jordan(b)))
}

var sub = function (a, b) {
    return from_jordan(jordan_add(to_jordan(a), to_jordan([b[0], P.subtract(b[1]).mod(P)])))
}

var negate = function (a) {
    return [a[0], P.subtract(a[1]).mod(P)];
}

var ecPoint = function(a) {
    return mul([X,Y], a);
}



//Extract Bitcoin Public Key using R, S and Z values.

//secp256k1 constants
var P = new BN('fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f', 16);  //2²⁵⁶ - 2³² - 2⁹ - 2⁸ - 2⁷ - 2⁶ - 2⁴ - 1
var N = new BN('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 16);
var ZERO = new BN(0);
var SEVEN = new BN(7);
var X = new BN('79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798', 16);
var Y = new BN('483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8', 16);

//other constants my functions use
var ONE = new BN(1);
var TWO = new BN(2);
var THREE = new BN(3);
var Pp1d4 = new BN('3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0c', 16);


- // http://2coin.org/index.html?txid=e12fee70d2b0b308129b9b12d6f95be25617fc4f98c56900034ce466b3113074
- // var R = новый BN ('00aa7f028c41eaa2d1e6b13ab7d6de74a162a9e38e782783c69f2c94c3b5a73687', 16);
- // var S = новый BN ('0737baa0630cda8c21f025eee499934a4625eb9ab506ad7c38778aa2968cdfee', 16);
- // var Z = новый BN ('5534ef558bca21f84873907834d12e35b238850454ec6c3c62929c59b34853d4', 16);
    
    
console.log("R                = " + R.toString(16));
console.log("S                = " + S.toString(16));
console.log("Z                = " + Z.toString(16));

console.log();

// y² = x³ + 7

var x = R;
console.log("x                = " + x.toString(16));

var ySquared = x.pow(THREE).add(SEVEN).mod(P);
console.log("y²               = " + ySquared.toString(16));
//there are two possible answers that produce a squared result, eg, 4 = 2² or -2². We only need one of them, doesn't matter which.

var y = ySquared.modPow(Pp1d4, P); //get the root of ySquared
//var y = P.minus(ySquared.modPow(Pp1d4, P)).mod(P); 
console.log("y                = " + y.toString(16));

console.log("-------------");
console.log();
console.log("2 ecPoints where x = R. I only need one of them, doesn't matter which");
console.log();
var ecPointK = [x, y];
console.log("ecPointK (1)     = " + ecPointK[0].toString(16) + ", " + ecPointK[1].toString(16));
console.log("ecPointK (2)     = " + ecPointK[0].toString(16) + ", " + P.minus(ecPointK[1]).mod(P).toString(16));

//pseudocode to get public key point
//pubKey  = (ecPoint(K) * (S/R)) - ecPoint(Z/R))

console.log();


console.log("-------------");
console.log();
console.log("calculate some more constants from the supplied R, S and Z values");
console.log();

var SdR = S.multiply(R.modInv(N)).mod(N);
console.log("S/R              = " + SdR.toString(16));

var ZdR = Z.multiply(R.modInv(N)).mod(N);
console.log("Z/R              = " + ZdR.toString(16));

var ecPointKmSdR = mul(ecPointK, SdR);
console.log("ecPointK * (S/R) = [" + ecPointKmSdR[0].toString(16) + ", " + ecPointKmSdR[1].toString(16) + "]");

var ecPointZdR = mul([X, Y], ZdR);
console.log("ecPoint(Z/R)     = [" + ecPointZdR[0].toString(16) + ", " + ecPointZdR[1].toString(16) + "]");

console.log();
console.log("-------------");
console.log();
console.log("2 Possible ecPoints for the public key. (depends on which ecPointK was used)");
console.log();
var ecPointPubKey1 = sub(ecPointKmSdR, ecPointZdR); 
var ecPointPubKey2 = sub(negate(ecPointKmSdR), ecPointZdR); //when using P - root of ySquared value


console.log("pubKeyPoint (1)  = [" + ecPointPubKey1[0].toString(16) + ", " + ecPointPubKey1[1].toString(16) + "]");
console.log("pubKeyPoint (2)  = [" + ecPointPubKey2[0].toString(16) + ", " + ecPointPubKey2[1].toString(16) + "]");
console.log();

console.log("-------------");
console.log();
console.log("4 possible Bitcoin formatted public keys. (depends on which ecPointK was used)");
console.log();

console.log("uncompressed pubkey (1)        = \n04" + ecPointPubKey1[0].toString(16) + ecPointPubKey1[1].toString(16));
if(ecPointPubKey1[1].mod(TWO).eq(ZERO)){
    console.log("compressed pubkey (1) (even y) = \n02" + ecPointPubKey1[0].toString(16));
}else{
    console.log("compressed pubkey (1) (odd y)  = \n03" + ecPointPubKey1[0].toString(16));
}
console.log();

console.log("uncompressed pubkey (2)        = \n04" + ecPointPubKey2[0].toString(16) + ecPointPubKey2[1].toString(16));
if(ecPointPubKey2[1].mod(TWO).eq(ZERO)){
    console.log("compressed pubkey (2) (even y) = \n02" + ecPointPubKey2[0].toString(16));
}else{
    console.log("compressed pubkey (2) (odd y)  = \n03" + ecPointPubKey2[0].toString(16));
}
console.log();

File: getPubKeyFromRSZ.py
def modInv(n, p):
    return pow(n, p - 2, p)


def jordan_isinf(p):
    return p[0][0] == 0 and p[1][0] == 0


def mulcoords(c1, c2):
    return (c1[0] * c2[0] % P, c1[1] * c2[1] % P)


def mul_by_const(c, v):
    return (c[0] * v % P, c[1])


def addcoords(c1, c2):
    return ((c1[0] * c2[1] + c2[0] * c1[1]) % P, c1[1] * c2[1] % P)


def subcoords(c1, c2):
    return ((c1[0] * c2[1] - c2[0] * c1[1]) % P, c1[1] * c2[1] % P)


def invcoords(c):
    return (c[1], c[0])


def jordan_add(a, b):
    if jordan_isinf(a):
        return b
    if jordan_isinf(b):
        return a
    if (a[0][0] * b[0][1] - b[0][0] * a[0][1]) % P == 0:
        if (a[1][0] * b[1][1] - b[1][0] * a[1][1]) % P == 0:
            return jordan_double(a)
        else:
            return ((0, 1), (0, 1))

    xdiff = subcoords(b[0], a[0])
    ydiff = subcoords(b[1], a[1])
    m = mulcoords(ydiff, invcoords(xdiff))
    x = subcoords(subcoords(mulcoords(m, m), a[0]), b[0])
    y = subcoords(mulcoords(m, subcoords(a[0], x)), a[1])
    return (x, y)


def jordan_double(a):
    if jordan_isinf(a):
        return ((0, 1), (0, 1))
    num = addcoords(mul_by_const(mulcoords(a[0], a[0]), 3), [0, 1])
    den = mul_by_const(a[1], 2)
    m = mulcoords(num, invcoords(den))
    x = subcoords(mulcoords(m, m), mul_by_const(a[0], 2))
    y = subcoords(mulcoords(m, subcoords(a[0], x)), a[1])
    return (x, y)


def jordan_multiply(a, n):
    if jordan_isinf(a) or n == 0:
        return ((0, 0), (0, 0))
    if n == 1:
        return a
    if n < 0 or n >= N:
        return jordan_multiply(a, n % N)
    if n % 2 == 0:
        return jordan_double(jordan_multiply(a, n // 2))
    else:  # n % 2 == 1:
        return jordan_add(jordan_double(jordan_multiply(a, n // 2)), a)


def to_jordan(p):
    return ((p[0], 1), (p[1], 1))


def from_jordan(p):
    return (p[0][0] * modInv(p[0][1], P) % P, p[1][0] * modInv(p[1][1], P) % P)


def mul(a, n):
    """
    Multiply an ECPoint.
    @param {number} a - An ECPoint
    @param {number} n - A Big Number
    """
    return from_jordan(jordan_multiply(to_jordan(a), n))


def div(a, n):
    """
    Divide an ECPoint.
    @param {number} a - An ECPoint
    @param {number} n - A Big Number
    """
    return from_jordan(jordan_multiply(to_jordan(a), modInv(n, N) % N))


def add(a, b):
    """
    Add two ECPoints.
    @param {number} a - An ECPoint
    @param {number} b - An ECPoint
    """
    return from_jordan(jordan_add(to_jordan(a), to_jordan(b)))


def sub(a, b):
    """
    Subtract two ECPoints.
    @param {number} a - An ECPoint
    @param {number} b - An ECPoint
    """
    return from_jordan(jordan_add(to_jordan(a), to_jordan((b[0], P - (b[1] % P)))))


def negate(a):
    return (a[0], P - (a[1] % P))


def ecPoint(a):
    return mul((X, Y), a)


# Extract Bitcoin Public Key using R, S and Z values.

# secp256k1 constants
# 2²⁵⁶ - 2³² - 2⁹ - 2⁸ - 2⁷ - 2⁶ - 2⁴ - 1
P = 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f
N = 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141
X = 0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798
Y = 0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8
Pp1d4 = 0x3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0c


# https://2xoin.com/tx/9ec4bc49e828d924af1d1029cacf709431abbde46d59554b62bc270e3b29c4b1
R = 0xd47ce4c025c35ec440bc81d99834a624875161a26bf56ef7fdc0f5d52f843ad1
S = 0x44e1ff2dfd8102cf7a47c21d5c9fd5701610d04953c6836596b4fe9dd2f53e3e
Z = 0xc0e2d0a89a348de88fda08211c70d1d7e52ccef2eb9459911bf977d587784c6e


# https://2xoin.com/tx/5cc5364a43ee9212387bfd45fa2e2c5e8ed7b2a32fa2f8e3084f6c8845ca3e15
# R = 0x3b78ce563f89a0ed9414f5aa28ad0d96d6795f9c63
# S = 0x5d38a230719d7282d6650c7a19ea810531501d391448aa61061828a073da404d
# Z = 0x422e18e3c79ed7ba9eb41754d26ab846d6c0644abb51eeca74a8973d4d8e656b

# see js version for more example RSZ's to try.


print("R                = 0x%x" % R)
print("S                = 0x%x" % S)
print("Z                = 0x%x" % Z)


print()

# y² = x³ + 7

x = R
print("x                = 0x%x" % x)

ySquared = ((x ** 3) + 7) % P
print("y²               = 0x%x" % ySquared)
# there are two possible answers that produce a squared result, eg, 4 = 2² or -2². We only need one of them, doesn't matter which.

y = pow(ySquared, Pp1d4, P)  # get the root of ySquared
#y = (P - pow(ySquared, Pp1d4, P)) % P
print("y                = 0x%x" % y)

print("-------------")
print()
print("2 ecPoints where x = R. I only need one of them, doesn't matter which")
print()

ecPointK = (x, y)
print("ecPointK (1)     = 0x%x, 0x%x" % (ecPointK[0], ecPointK[1]))
print("ecPointK (2)     = 0x%x, 0x%x" % (ecPointK[0], (P - (ecPointK[1] % P))))

# pseudocode to get public key point
# pubKey = (ecPoint(K) * (S/R)) - ecPoint(Z/R))

print()


print("-------------")
print()
print("calculate some more constants from the supplied R, S and Z values")
print()

SdR = S * modInv(R, N) % N
print("S/R              = 0x%x" % SdR)

ZdR = Z * modInv(R, N) % N
print("Z/R              = 0x%x" % ZdR)


ecPointKmSdR = mul(ecPointK, SdR)
print("ecPointK * (S/R) = (0x%x, 0x%x)" % (ecPointKmSdR[0], ecPointKmSdR[1]))
ecPointZdR = mul([X, Y], ZdR)
print("ecPoint(Z/R)     = (0x%x, 0x%x)" % (ecPointZdR[0], ecPointZdR[1]))

print()
print("-------------")
print()
print("2 Possible ecPoints for the public key. (depends on which ecPointK was used)")
print()
ecPointPubKey1 = sub(ecPointKmSdR, ecPointZdR)
# when using P - root of ySquared value
ecPointPubKey2 = sub(negate(ecPointKmSdR), ecPointZdR)


print("pubKeyPoint (1)  = (0x%x, 0x%x)" %
      (ecPointPubKey1[0], ecPointPubKey1[1]))
print("pubKeyPoint (2)  = (0x%x, 0x%x)" %
      (ecPointPubKey2[0], ecPointPubKey2[1]))
print()

print("-------------")
print()
print("4 possible Bitcoin formatted public keys. (depends on which ecPointK was used)")
print()

print("uncompressed pubkey (1)        = \n04%x%x" %
      (ecPointPubKey1[0], ecPointPubKey1[1]))
if ecPointPubKey1[1] % 2 == 0:
    print("compressed pubkey (1) (even y) = \n02%x" % (ecPointPubKey1[0]))
else:
    print("compressed pubkey (1) (odd y)  = \n03%x" % (ecPointPubKey1[0]))

print()

print("uncompressed pubkey (2)        = \n04%x%x" %
      (ecPointPubKey2[0], ecPointPubKey2[1]))
if ecPointPubKey2[1] % 2 == 0:
    print("compressed pubkey (2) (even y) = \n02%x" % (ecPointPubKey2[0]))
else:
    print("compressed pubkey (2) (odd y)  = \n03%x" % (ecPointPubKey2[0]))

print()


# simple test
#  1st i create an ecpoint of 10 (secp256k1)
#myECPoint = mul((x, y), 10)
# console.log("myECPoint = [" + myECPoint[0].toString(16) + \
#             ", " + myECPoint[1].toString(16) + "]")
# // next i will divide it by 5. must convert 5 to a bignumber
#ecResult = div(myECPoint, 5)
# console.log("ecResult = [" + ecResult[0].toString(16) + \
#             ", " + ecResult[1].toString(16) + "]")
# // //result should equal ecpoint of 2 (secp256k1)
# ecTest = mul((x, y), 2)  # expected result
#print(ecResult[0] == ecTest[0] and ecResult[1] == ecTest[1])
